### PR TITLE
forgejo: refactor to support multiple versions, 7.0.5 -> 8.0.0, forgejo-lts: init at 7.0.6

### PR DESCRIPTION
**Pull request by emilylange**
## Description of changes

https://codeberg.org/forgejo/forgejo/releases/tag/v7.0.6

https://codeberg.org/forgejo/forgejo/releases/tag/v8.0.0

The general structure is copied from `incus` and `incus-lts`.
@adamcstephens messaged me on Matrix shortly after the release announcement to coordinate, and we settled on something like this.

Note that @adamcstephens did not take a look at this yet, so whoever reviews this, don't consider this to be set in stone or anything :)

Please refer to the commit messages for details.

Edit: It's getting late for me, so it's probably for the better not to spend a bunch on time on release notes just yet.
I suppose we work on them once everyone agreed on the new structure and defaults. Maybe even in a follow-up PR^^

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by adamcstephens** [comment]
New files need to be formatted now.

File: nixos/modules/services/misc/forgejo.nix [hunk 1]
```diff
@@ -66,7 +66,7 @@ in
     services.forgejo = {
       enable = mkEnableOption "Forgejo, a software forge";
 
-      package = mkPackageOption pkgs "forgejo" { };
+      package = mkPackageOption pkgs "forgejo-lts" { };
```
Comment (ambroisie) [inline nixos/modules/services/misc/forgejo.nix:69]:
```suggestion
      package = mkPackageOption pkgs "Forgejo" { default = [ "forgejo-lts" ]; };
```

**Review by ambroisie — COMMENTED** [review]
My one question is whether we should default to `forgejo-lts` always, or have something similar to Nextcloud:

* each package is named explicitly with its major version (`forgejo_7`, `forgejo_8`, potentially an alias from `forgejo_lts` to `forgejo_7`).
* set the `package` default depending on `system.stateVersion`, with warnings when the package is "too old".

**Review by bendlas — APPROVED** [review]
Tested both 7.0.6 and 8.0.0

- both seem to work
- upgrade from 7.0.6 to 8.0.0 works

**Comment by emilylange** [comment]
> New files need to be formatted now.

The file [it's complaining about](https://github.com/NixOS/nixpkgs/actions/runs/10172939034/job/28136256440?pr=331187) is `generic.nix`, a file that has been renamed and isn't new.
I believe this is either a regression or a false positive.
@infinisil can you please comment on that?

```
Ignoring file nixos/modules/services/misc/forgejo.nix because it's not formatted in the base commit
Ignoring file nixos/tests/all-tests.nix because it's not formatted in the base commit
Ignoring file nixos/tests/forgejo.nix because it's not formatted in the base commit
pkgs/by-name/fo/forgejo/generic.nix: not formatted
Ignoring file pkgs/by-name/fo/forgejo/package.nix because it's not formatted in the base commit
Ignoring file pkgs/top-level/all-packages.nix because it's not formatted in the base commit
Some new/changed Nix files are not properly formatted
Please run the following in `nix-shell`:
nixfmt 'pkgs/by-name/fo/forgejo/generic.nix'
Error: Process completed with exit code 1.
```

---
@ambroisie

> My one question is whether we should default to `forgejo-lts` always, or have something similar to Nextcloud:
> 
>     * each package is named explicitly with its major version (`forgejo_7`, `forgejo_8`, potentially an alias from `forgejo_lts` to `forgejo_7`).
> 
>     * set the `package` default depending on `system.stateVersion`, with warnings when the package is "too old".

Unlike Nextcloud, Forgejo does not have overlapping releases, so I am not quite sure what would be the benefit of doing it that way.
See <https://forgejo.org/docs/latest/developer/release/>.
Would you mind commenting on the benefits when adopting your suggested approach?

**Review by mweinelt — APPROVED** [review]
Looks good to me as is. Thank you!

**Comment by ambroisie** [comment]
> Would you mind commenting on the benefits when adopting your suggested approach?

Simply because it's the only case I know off the top of my head of a module caring about the version of its software, so I thought "why not align with the way it's done over there". Your justification sounds good to me.

**Comment by tnias** [comment]
I rebased the changes on nixos-unstable (52ec9ac3b12395ad677e8b62106f0b98c1f8569d) and used that to migrated a setup from gitea 1.22.1 to forgejo 8.0.0. Everything seems to be running fine. Thanks :)

**Comment by adamcstephens** [comment]
> The file [it's complaining about](https://github.com/NixOS/nixpkgs/actions/runs/10172939034/job/28136256440?pr=331187) is `generic.nix`, a file that has been renamed and isn't new.

At least according to the github view, generic.nix is considered a new file and package.nix is just a modification. I'd say just go ahead and format it in that case since history is likely lost anyway.

**Review by adamcstephens — APPROVED** [review]

**Comment by infinisil** [comment]
> > The file [it's complaining about](https://github.com/NixOS/nixpkgs/actions/runs/10172939034/job/28136256440?pr=331187) is `generic.nix`, a file that has been renamed and isn't new.
> 
> At least according to the github view, generic.nix is considered a new file and package.nix is just a modification. I'd say just go ahead and format it in that case since history is likely lost anyway.

Yeah, if Git can't detect the rename, there's no way to distinguish it from a new file, reformatting seems good :)

**Pull request merged**